File: arch/src/main/java/com/github/lany192/box/fragment/ModelFragment.java
package com.github.lany192.box.fragment;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;

import com.github.lany192.box.mvvm.LifecycleViewModel;

public abstract class ModelFragment extends BasicFragment {

    protected <T extends LifecycleViewModel> T getFragmentViewModel(@NonNull Class<T> modelClass) {
        T viewModel =  new ViewModelProvider(this).get(modelClass);
        getLifecycle().addObserver(viewModel);
        return viewModel;
    }

    protected <T extends LifecycleViewModel> T getActivityViewModel(@NonNull Class<T> modelClass) {
        T viewModel =  new ViewModelProvider(requireActivity()).get(modelClass);
        getLifecycle().addObserver(viewModel);
        return viewModel;
    }
}

File: arch/src/main/java/com/github/lany192/arch/fragment/ModelFragment.kt
package com.github.lany192.arch.fragment

import androidx.annotation.CallSuper
import androidx.lifecycle.ViewModel
import androidx.viewbinding.ViewBinding
import java.lang.reflect.ParameterizedType

/**
 * ViewBinding实现基类
 */
abstract class ModelFragment<VM : ViewModel, VB : ViewBinding> : BindingFragment<VB>() {
    lateinit var viewModel: VM

    @CallSuper
    override fun init() {
        val clazz =
            (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<VM>
        viewModel = createDefaultViewModel(clazz)
    }

    protected fun createDefaultViewModel(clazz: Class<VM>): VM {
        return getViewModel(clazz)
    }
}